### PR TITLE
Add weight normalization implementation

### PR DESCRIPTION
**Pull request by colesbury**
This adds forward "pre-hooks" which get called before the module's
forward() method. Weight norm is implemented as a hook which calculates
the weight variable from the weight_g and weight_v every iteration.

Based on @rtqichen implementation.

### CODE REVIEW

**Comment by colesbury** [comment]
cc @rtqichen

I went with an implementation based on hooks, which is similar to your original function decorator implementation, but can be pickled and works with DataParallel.

File: torch/nn/utils/weight_norm.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    See https://arxiv.org/abs/1602.07868
+
+    Args:
+        module (nn.Module): containing module
+        name (str, optional): name of weight parameter
+        dim (int, optional): output dimension of weight parameter
+
+    Example::
+
+        >>> m = weight_norm(nn.Linear(20, 40))
```
Comment (soumith) [inline torch/nn/utils/weight_norm.py]: it's not obvious from the example that `name='weight'` is being used here.
Can you change the example in two ways:
- `m = weight_norm(nn.Linear(20, 40), 'weight')
- show how to weight_norm both weight and bias of the nn.Linear (is it separate calls? or does name take a tuple of strings too?

File: torch/nn/utils/weight_norm.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    Weight normalization is a reparameterization that decouples the magnitude
+    of a weight tensor from its direction. This replaces the parameter specified
+    by `name` (e.g. "weight") with two parameters: one specifying the magnitude
+    (e.g. "weight_g") and one specifying the direction (e.g. "weight_v").
+    The module's forward function is wrapped to first recompute the weight
+    tensor from the magnitude and direction.
+
+    The magnitude parameter ("weight_g") contains weight.size(dim) elements.
+    By default, weight is normalized per activation in an `nn.Linear` layer and
+    per output-plane in a convolutional layer. If `dim` is `None`, the entire
```
Comment (soumith) [inline torch/nn/utils/weight_norm.py]: > If `dim` is `None`, the entire weight tensor is normalized.

Then why is `dim`'s default value `0`?

**Review by soumith — CHANGES_REQUESTED** [review]

File: torch/nn/utils/weight_norm.py (hunk 2 above)
Comment (colesbury) [inline torch/nn/utils/weight_norm.py]: Because by default, each output channel is normalized separately. (This is how its specified in the paper and how it's implemented in nn.WeightNorm). In most of our module's this corresponds to dim `0`.

File: torch/nn/utils/weight_norm.py (hunk 1 above)
Comment (colesbury) [inline torch/nn/utils/weight_norm.py]: Why would you weight norm bias...?

Comment (soumith) [inline torch/nn/utils/weight_norm.py]: it's not really to show how to weight norm bias. Right now I'm not sure what the returned object is out of weight_norm. If i have two weights (like bilinear), then do I `m=weight_norm(m, 'weight1'); m=weight_norm(m, 'weight2')` ?
Maybe a simpler way to refine the example is to print out the type of `m`, which will provide some clarity.

Comment (colesbury) [inline torch/nn/utils/weight_norm.py]: I documented the return type. I'm not a huge fan of adding examples of weight normalizing something like bias (unless someone has showed that it actually makes sense). (Bilinear takes in two inputs, but only has one weight tensor)

If you have two parameters, you need two calls to `weight_norm`. Either:

```
m = weight_norm(weight_norm(m, 'weight1'), 'weight2')
```

or
```
m = ...
weight_norm(m, 'weight1')
weight_norm(m, 'weight2')
```

Like most of our other methods (e.g. `.cuda()`) on modules, `weight_norm` returns the passed in module as a convenience.

**Review by soumith — APPROVED** [review]

**Comment by hanzhanggit** [comment]
@rtqichen @colesbury I did not find the data dependent initialization(which is used in the original paper) for the weight norm. I thought it was important to the weight norm. Do we plan to have it implemented or any examples about it?

**Comment by xwuaustin** [comment]
@rtqichen  Thanks for contribution for this code. 
@hanzhanggit @soumith @colesbury @all  
Hi everyone who read this post. I have some questions regarding to weight_norm. It would be great if you can help. 
I tried to implement the weight_norm for each convolution and linear layer (check the code here https://github.com/xwuaustin/weight_norm/blob/master/cifar10_tutorial_weightNorm.py ).  However, the training loss in CIFAR-10 seems no difference to the original setting (see the picture below) at the first 2 epochs.  


Now questions:
### 1. Is there something wrong with the code I modified? I used the code from cifar10_tutorial in pytorch. All I did is to add the wieghtNorm at each layer.

import torch.nn.utils.weight_norm as weightNorm
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        ### we use weight normalization after each convolutions and linear transfrom
        **self.conv1 = weightNorm(nn.Conv2d(3, 6, 5),name = "weight")**
        #print (self.conv1._parameters.keys())
        self.pool = nn.MaxPool2d(2, 2)
        **self.conv2 =weightNorm(nn.Conv2d(6, 16, 5),name = "weight")**
        **self.fc1 = weightNorm(nn.Linear(16 * 5 * 5, 120),name = "weight")**
        **self.fc2 = weightNorm(nn.Linear(120, 84),name = "weight")**
        **self.fc3 = weightNorm(nn.Linear(84, 10),name = "weight")**


### 2 Is the update of the weights and bias, namely 'weight_g', 'weight_v',  using the formulation:


### 3. Can we do the initialization as the paper suggested? 


Thanks. Looking for your responds. :)

**Comment by sajadn** [comment]
I wrote a function which does the data-dependent initialization for weight norm layers (now it's just for conv layers, but it's straight forward to adapt it). 
https://gist.github.com/sajadn/f29b14cd52023b614c334708c2a4345f

**Comment by tfjgeorge** [comment]
@colesbury Hi! Could you kindly explain what is the advantage (if any) of implementing weight norm with hooks as you do, versus implementing it as a new layer type (nn.Module) ?

**Comment by colesbury** [comment]
@tfjgeorge An nn.Module probably would made sense, but I don't remember the details so I'm not sure. (It's been over four years)
